### PR TITLE
Add QR code scanner

### PR DESCRIPTION
**Pull request by bgoncal**
## Summary

Add QR code scanner that can be triggered from HA and receive result response
## Screenshots

  

https://github.com/home-assistant/iOS/assets/5808343/0e4d4f18-7507-44f2-8844-e701448943dd

## Link to pull request in Documentation repository

Documentation: home-assistant/companion.home-assistant#

### CODE REVIEW

**Comment by bgoncal** [comment]
https://github.com/home-assistant/frontend/pull/19570

File: Sources/App/AppDelegate.swift [hunk 1]
```diff
@@ -260,6 +262,13 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         IntentHandlerFactory.handler(for: intent)
     }
 
+    func application(
+        _ application: UIApplication,
+        supportedInterfaceOrientationsFor window: UIWindow?
```
Comment (zacwest) [inline Sources/App/AppDelegate.swift]: you should not need to do this. use UIViewController-based interface orientation config.

File: Sources/App/QRCodeScanner/BarcodeScannerView.swift [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            DispatchQueue.main.async {
+                AppDelegate.orientationLock = UIInterfaceOrientationMask.all
+                UINavigationController.attemptRotationToDeviceOrientation()
+            }
+        }
+    }
+
+    private var topInformation: some View {
+        VStack(spacing: 8) {
+            Button(action: {
```
Comment (zacwest) [inline Sources/App/QRCodeScanner/BarcodeScannerView.swift:52]: feelsl ike this needs an accessibility title

File: Sources/App/QRCodeScanner/Camera/BarcodeScannerCamera.swift [hunk 3]
```diff
@@ -0,0 +1,335 @@
+import AVFoundation
+import CoreImage
+import Shared
+import UIKit
+
+class BarcodeScannerCamera: NSObject {
```
Comment (zacwest) [inline Sources/App/QRCodeScanner/Camera/BarcodeScannerCamera.swift:6]: can you describe what this does and why it's a better solution than e.g. DataScannerViewController?

File: Sources/App/QRCodeScanner/Camera/BarcodeScannerCamera.swift [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+    func metadataOutput(
+        _ output: AVCaptureMetadataOutput,
+        didOutput metadataObjects: [AVMetadataObject],
+        from connection: AVCaptureConnection
+    ) {
+        if let metadataObject = metadataObjects.first {
+            let format = metadataObject.type.haString
+            guard let readableObject = metadataObject as? AVMetadataMachineReadableCodeObject else { return }
+            guard let stringValue = readableObject.stringValue else { return }
+            AudioServicesPlaySystemSound(SystemSoundID(kSystemSoundID_Vibrate))
```
Comment (zacwest) [inline Sources/App/QRCodeScanner/Camera/BarcodeScannerCamera.swift]: why this over haptic engine's success haptic?

**Review by zacwest — CHANGES_REQUESTED** [review]

File: Sources/App/QRCodeScanner/Camera/BarcodeScannerCamera.swift (hunk 4 above)
Comment (bgoncal) [reply Sources/App/QRCodeScanner/Camera/BarcodeScannerCamera.swift]: I was having an issue where the haptics were "glitching" while the view controller dismisses, difficult to explain in writting, but basically it was vibrating severald times while dismissing.
It endup being my test device, after rebooting it it's good now. I have updated the implementation

File: Sources/App/QRCodeScanner/BarcodeScannerView.swift (hunk 2 above)
Comment (bgoncal) [reply Sources/App/QRCodeScanner/BarcodeScannerView.swift:52]: Good catch

File: Sources/App/QRCodeScanner/Camera/BarcodeScannerCamera.swift (hunk 3 above)
Comment (bgoncal) [reply Sources/App/QRCodeScanner/Camera/BarcodeScannerCamera.swift:6]: DataScannerViewController is iOS16+

**Pull request merged**